polygeist/lib/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.cpp

```diff
@@ -1075,7 +1060,46 @@ OffsetVector::getConstIntegerValue(size_t row, DataFlowSolver &solver) const {
 // MemoryAccess
 //===----------------------------------------------------------------------===//
 
+[[maybe_unused]] raw_ostream &
+mlir::polygeist::operator<<(raw_ostream &os, const MemoryAccess &access) {
+  os << "--- MemoryAccess ---\n\n";
+  os << "AccessMatrix:\n" << access.getAccessMatrix() << "\n";
+  os << "OffsetVector:\n" << access.getOffsetVector() << "\n";
+  os << "\n------------------\n";
+  return os;
+}
+
+MemoryAccessMatrix
+MemoryAccess::getIntraThreadAccessMatrix(unsigned numThreads) const {
+  assert(numThreads < matrix.getNumColumns() &&
+         "Expecting 'numThreadsColumns' to be smaller than the number of "
+         "columns in the memory access matrix");
+  std::vector<size_t> v(numThreads);
+  std::iota(v.begin(), v.end(), 0);
+  std::set<size_t> columns(v.begin(), v.end());
+  return matrix.getColumns(columns);
+}
+
+MemoryAccessMatrix
+MemoryAccess::getInterThreadAccessMatrix(unsigned numThreads) const {
+  assert(numThreads < matrix.getNumColumns() &&
+         "Expecting 'numThreadsColumns' to be smaller than the number of "
+         "columns in the memory access matrix");
+  std::vector<size_t> v(matrix.getNumColumns() - numThreads);
+  std::iota(v.begin(), v.end(), numThreads);
+  std::set<size_t> columns(v.begin(), v.end());
+  return matrix.getColumns(columns);
+}
+
 MemoryAccessPattern MemoryAccess::classify(DataFlowSolver &solver) const {
+  return MemoryAccess::classify(matrix, offsets, solver);
+}
+
+MemoryAccessPattern MemoryAccess::classify(const MemoryAccessMatrix &matrix,
+                                           const OffsetVector &offsets,
+                                           DataFlowSolver &solver) {
+  LLVM_DEBUG(llvm::dbgs() << "inter-thread matrix:\n" << matrix << "\n");
```

Comment:
How do we know `matrix` is the inter-thread matrix here?

Reply:
done